File: DSA/LakshaKundra_590013237/Day_76_Question_1.py
class Node:
    def __init__(self, key):
        self.data = key
        self.left = None
        self.right = None

def to_sum_tree(node):
    if not node:
        return 0
    left_sum = to_sum_tree(node.left)
    right_sum = to_sum_tree(node.right)
    old_val = node.data
    node.data = left_sum + right_sum
    return node.data + old_val

def preorder(node):
    if not node:
        return
    print(node.data, end=" ")
    preorder(node.left)
    preorder(node.right)

# Example 1
root = Node(10)
root.left = Node(-2)
root.right = Node(6)
root.left.left = Node(8)
root.left.right = Node(-4)
root.right.left = Node(7)
root.right.right = Node(5)

to_sum_tree(root)
preorder(root)  # Output: 20 4 0 0 12 0 0

print()

# Example 2
root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.right = Node(5)

to_sum_tree(root)
preorder(root)  # Output: 14 9 0 0 0 3 0

File: DSA/RATNA_KHANNA_590013141/DAY_76_QUESTION_1.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* left;
    struct Node* right;
};

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->left = newNode->right = NULL;
    return newNode;
}

int toSumTree(struct Node* root) {
    if (root == NULL)
        return 0;

    int leftSum = toSumTree(root->left);
    int rightSum = toSumTree(root->right);

    int oldValue = root->data;
    root->data = leftSum + rightSum;

    return oldValue + root->data;
}

void printTree(struct Node* root) {
    if (root == NULL)
        return;
    printTree(root->left);
    printf("%d ", root->data);
    printTree(root->right);
}

int main() {
    int n;
    printf("Enter the number of nodes: ");
    scanf("%d", &n);

    struct Node* nodes[n];
    printf("Enter the nodes data:\n");
    for (int i = 0; i < n; i++) {
        int val;
        scanf("%d", &val);
        nodes[i] = createNode(val);
    }

    // Manually creating the tree as per the example
    nodes[0]->left = nodes[1];
    nodes[0]->right = nodes[2];
    nodes[1]->left = nodes[3];
    nodes[2]->left = nodes[4];
    nodes[2]->right = nodes[5];

    printf("Original tree in inorder traversal:\n");
    printTree(nodes[0]);
    printf("\n");

    toSumTree(nodes[0]);

    printf("Sum tree in inorder traversal:\n");
    printTree(nodes[0]);
    printf("\n");

    return 0;
}
